### ISSUE TITLE
Is it possible to use marching cubes mesh extraction for xyz point cloud?

### ISSUE THREAD

**Issue by 1939938853** [intel-isl/Open3D#2365]
Hi,

Instead of [RGBD](http://www.open3d.org/docs/release/tutorial/Advanced/rgbd_integration.html), just wonder it is possible to use marching cubes mesh extraction for xyz point cloud?

Thanks

**Comment by griegler** [comment]
No, you need a (T)SDF volume to use marching cubes and you can not simply turn a point cloud into a (T)SDF volume.

**Comment by 1939938853** [comment]
From [here](https://github.com/PointCloudLibrary/pcl/blob/master/tools/marching_cubes_reconstruction.cpp), it looks that PCL can use Marching Cubes to create mesh?

**Comment by griegler** [comment]
Yes, you can turn a point cloud with associated normals into a (T)SDF volume. Unfortunately, that is not implemented in Open3D yet. @benjaminum what do you think, how much work this would be?

**Comment by benjaminum** [comment]
I had a quick look at the TSDF volume code, which is very specific to depth images. This means we need to implement this from scratch. Adding an efficient implementation can take some time. We should discuss if we can integrate this with the new Tensor-based data structures. Maybe we can also include @theNded in the discussion.

**Comment by 1939938853** [comment]
Marching Cubes based surface reconstruction seems promising in the aspect of real-time related performance.

-  Can avoid normal estimation? Normal estimation is not so "cheap", also GPU acceleration is not so effective, especially when the point cloud is not so big.  Furthermore to make all normal direction towards outside is not so "cheap" either.  

- From [Nvidia's](https://www.nvidia.com/content/gtc-2010/pdfs/gtc10_session2020_hp5mc_dyken_ziegler.pdf)  early presentation, it seems that Marching Cubes is parallelization friendly?

Surface reconstruction from point cloud, except meshing, perhaps the other better  choice is the [surface splatting](https://github.com/sebastianlipponer/surface_splatting)? I guess that the Open3D visualization of point cloud with normal has some kind of splatting?  But it is not "water-tight". If Open3D can also realize the surface splatting as shown in the this [link](https://github.com/sebastianlipponer/surface_splatting) , that could be great. Even though surface splatting require per-calculated normal, but seems that it is not sensitive to the normal direction.

**Comment by benjaminum** [comment]
Surface splatting and marching cubes are very different.
Does you point cloud describe a surface and you just want to visualize this surface?

**Comment by 1939938853** [comment]
For me, both are interesting

-  For further surface (mesh) based processing, such as registration,  a mesh surface may be preferred. In this case less points (vertices) can be chosen  in order to speedup the process. For real-time processing, the key point is that the processing must be fast and robust enough.

-  On the other hand, with the power of point cloud processing provided by Open3D, many 3D processing can be conducted in  point-domain (space), people just need good final result visualization or rendering. In this case water-tight splatting may be is a good choice?

**Comment by benjaminum** [comment]
Do you have some data that you could show? Seeing the data might help to better understand what Open3D is lacking.
I am also a bit confused by the _real-time_ processing part which can be very difficult to achieve for many problems.

**Comment by 1939938853** [comment]
Hello benjaminum,

Thank you for your replay!

To accelerate 3D data processing is really not so easy. Yesterday I tested [PCL's marching cubes](https://github.com/PointCloudLibrary/pcl/blob/master/tools/marching_cubes_reconstruction.cpp), when point size is more than 10 thousands, depending on the grid size, the processing may take several hours. Not acceptable.

Open3D contributors have already done an excellent work, in API simplification, in  speedup data processing and so on.  Hopefully in the upcoming release, you have more breakthrough, especially in parallelization and GPU acceleration on both Windows and Linux. 

I'm not good at the underneath algorithm. My feeling is that many calculations is based on the kdtree and the kdtree is not so parallelization friendly.  From Nvidia I saw the [HGMM](https://developer.download.nvidia.com/video/gputechconf/gtc/2019/presentation/s9623-gpu-accelerated-3d-point-cloud-processing-with-hierarchical-gaussian-mixtures.pdf), might be some part calculation can use HGMM?

Talking about real-time and 3D data. To me

For visualization

- to keep rather detailed surface information, point size at least  needs to be 10-20 thousands.

- to show rather smooth and continuous 3D data streaming, I think at least the frame rate should be > 10 fps

- you also need to have some other processing such as outlier  removing, smoothing, registration etc to be involved.

For control, such as in-house and out-house navigation, that can be another scenario. points can be fewer, may not need surface, but may need object detection, frame rate may need to be much higher.

Anyway, thank you for your attention and looking forward to the upcoming Open3D release!
